### PR TITLE
pools: Fix problem that Err() must not be called before Done was closed.

### PR DESCRIPTION
**Pull request by michael-berlin**
@sougou

### CODE REVIEW

**Comment by enisoc** [comment]
LGTM

File: go/pools/resource_pool.go [hunk 1]
```diff
@@ -104,9 +104,12 @@ func (rp *ResourcePool) TryGet() (resource Resource, err error) {
 
 func (rp *ResourcePool) get(ctx context.Context, wait bool) (resource Resource, err error) {
 	// If ctx has already expired, avoid racing with rp's resource channel.
-	if ctx.Err() != nil {
+	select {
+	case <-ctx.Done():
```
Comment (yaoshengzhe) [inline go/pools/resource_pool.go]: The intention is to test whether ctx is canceled or not. Based on following doc, ctx.Err() will return Canceled if ctx has already expired. In other words, we want to return ErrTimeout once deadline is reached (even if ctx is not Done).

  // Err returns a non-nil error value after Done is closed.  Err returns
    // Canceled if the context was canceled or DeadlineExceeded if the
    // context's deadline passed.  No other values for Err are defined.
    // After Done is closed, successive calls to Err return the same value.
    Err() error

Comment (yaoshengzhe) [inline go/pools/resource_pool.go]: I feel the current code does the right thing. @sougou

Comment (enisoc) [inline go/pools/resource_pool.go]: The open-source Context docs don't seem to mention it, but the return value of Err() is unspecified if you call it before the context's Done() channel is closed.

Comment (yaoshengzhe) [inline go/pools/resource_pool.go]: https://github.com/golang/net/blob/master/context/context.go#L91

ok, but then the code need to check ctx.Err() and then return nil, ErrTimeout, right ?

Comment (yaoshengzhe) [inline go/pools/resource_pool.go]: https://github.com/golang/net/blob/master/context/context.go#L91

After a closer look, I am still not sure. It seems you are right.

Comment (michael-berlin) [inline go/pools/resource_pool.go]: @sougou actually authored this fix ;-) But I was too lazy to specify this in the git commit. I'll do that the next time :)

To sum up, the change is fine with @sougou and I'll merge it now.

Comment (yaoshengzhe) [inline go/pools/resource_pool.go]: LTGM, go for it.

**Pull request merged**